mission/functions/systems/traits/fn_settrait.sqf

```diff
@@ -24,88 +25,61 @@ private _trait_config = (missionConfigFile >> "gamemode" >> "traits" >> _trait);
 	{
-		//We need these to set the isCustom flag on setUnitTrait!
-		private _vanilla_traits = [ 
-			"audibleCoef", 
-			"camouflageCoef", 
-			"loadCoef", 
-			"engineer", 
-			"explosiveSpecialist", 
-			"medic", 
-			"UAVHacker" 
-		];
-
-		// only allow one trait per player, if one is set then remove existing
-		_already_set = (_player_current_trait isNotEqualTo "");
-		if (_already_set) then
-		{
-			private _existing_traits = _current_traits_map get _player_current_trait;
-
-			// remove existing user from other set trait
-			private _to_remove = [];
-			{
-				if (_player isEqualTo _x) then
-				{
-					_to_remove pushBack _forEachIndex;
-				}
-			} forEach _existing_traits;
-
-			{
-				_existing_traits deleteAt _x
-			} forEach _to_remove;
-			// set trait to false
-			[_player,[_player_current_trait, false, !(_player_current_trait in _vanilla_traits)]] remoteExecCall ["setUnitTrait",_player];
-			_player setVariable ["vn_mf_dyn_trait_set", ""];
-		};
-
-		// set trait only if different from last
-		if (_trait isNotEqualTo _player_current_trait) then
-		{
-			_selected_traits pushBackUnique _player;
-            _current_traits_map set [_trait,_selected_traits];
-
-			// Do checks for tutorial for trait
-			["tookTraining", [_player, _trait]] call para_g_fnc_event_dispatch;
-
-			// mark player as already having set a trait
-			_player setVariable ["vn_mf_dyn_trait_set", _trait];
 
-			[_player,[_trait, true, !(_trait in _vanilla_traits)]] remoteExecCall ["setUnitTrait",_player];
-			{["TrainingSucceeded"] call para_c_fnc_show_notification} remoteExecCall ["call",_player];
+		switch (true) do {
+			// no player trait currently set -- add the new trait
+			// special case, needs to come first otherwise gets processed
+			// by the new_trait != current_trait block
+			case (_current_trait isEqualTo "") : {
+				[_trait, _player] call vn_mf_fnc_traits_player_add_trait;
+				[_trait, _player] call vn_mf_fnc_traits_db_push;
+				["TrainingSucceeded"] remoteExecCall ["para_c_fnc_show_notification", _player];
+			};
+			// a different trait was selected -- swap the assigned trait
+			case (_trait isNotEqualTo _current_trait) : {
+				[_current_trait, _player] call vn_mf_fnc_traits_player_remove_trait;
+				[_current_trait, _player] call vn_mf_fnc_traits_db_pop;
+				[_trait, _player] call vn_mf_fnc_traits_player_add_trait;
+				[_trait, _player] call vn_mf_fnc_traits_db_push;
+				["TrainingSwapped"] remoteExecCall ["para_c_fnc_show_notification", _player];
+			};
+			// the same trait selected -- remove the existing trait
+			case (_trait isEqualTo _current_trait) : {
+				[_trait, _player] call vn_mf_fnc_traits_player_remove_trait;
+				[_trait, _player] call vn_mf_fnc_traits_db_pop;
+				["TrainingRemoved"] remoteExecCall ["para_c_fnc_show_notification", _player];
+			};
 		};
-
-		// save changes to array and push to all players
-		missionNamespace setVariable ["vn_mf_traits_map", _current_traits_map];
-		publicVariable "vn_mf_traits_map";
```

Comment:
if player was not allowed to take a role, changes after pruning disconnected players would not have been broadcast here.